### PR TITLE
Feature/metrics sidepanel update

### LINKED ISSUE #632: [FEATURE] Metrics SelectedMetrics Panel
**Is your feature request related to a problem?**
Implement Metrics 2.9 UX - "Selected Metrics" left-panel

image

**Do you have any additional context?**
In this implementation, a metric from "Available Metrics" will be added to the "Selected Metrics" with a single-click.

### PR DESCRIPTION
**Pull request by pjfitzgibbons**
### Description
Update UX on Sidebar
- Move SearchBar into Sidebar
- show icons per-datasource
- compress metric-names (prometheus == "p..")

https://github.com/opensearch-project/dashboards-observability/assets/11318/c2735849-5eef-405a-a691-61c6595d24ae

### Issues Resolved
#632 - Partial Resolve (Subsequent PR will have Datasource selector)

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/components/custom_panels/helpers/utils.tsx [hunk 1]
```diff
@@ -175,6 +174,7 @@ export const fetchVisualizationById = async (
 
   await SavedObjectsActions.get({ objectId: savedVisualizationId })
     .then((res) => {
+      console.log({ res });
```
Comment (joshuali925) [inline public/components/custom_panels/helpers/utils.tsx:177]: remove debugging logs?

File: public/components/metrics/redux/slices/metrics_slice.ts [hunk 2]
```diff
@@ -3,71 +3,106 @@
...
-  const { http, pplService } = services;
-  const customData = await fetchCustomMetrics(http);
-  const remoteData = await fetchRemoteMetrics(pplService);
+export const loadMetrics = () => async (dispatch) => {
+  const { http, pplService } = coreRefs;
+  const customDataRequest = fetchCustomMetrics(http);
+  const remoteDataSourcesResponse = await pplServiceRequestor(pplService!, PPL_DATASOURCES_REQUEST);
+  const remoteDataSources = remoteDataSourcesResponse.data.DATASOURCE_NAME;
+
+  console.log('loadMetrics', { remoteDataSources });
```
Comment (joshuali925) [inline public/components/metrics/redux/slices/metrics_slice.ts:54]: same here

File: public/components/metrics/redux/slices/metrics_slice.ts [hunk 3]
```diff
@@ -3,71 +3,106 @@
...
+  dispatch(setDataSources(remoteDataSources));
+  dispatch(setDataSourceTitles(remoteDataSources));
+  dispatch(
+    setDataSourceIcons(coloredIconsFrom([OBSERVABILITY_CUSTOM_METRIC, ...remoteDataSources]))
+  );
 
-  return Promise.all([customData, ...remoteData]).then((datasets) => datasets.flat());
-});
+  const remoteDataRequests = await fetchRemoteMetrics(remoteDataSources);
+  console.log({ remoteDataRequests });
```
Comment (joshuali925) [inline public/components/metrics/redux/slices/metrics_slice.ts:62]: same here

File: public/components/metrics/sidebar/metric_name.tsx [hunk 4]
```diff
@@ -14,30 +25,19 @@ interface IMetricNameProps {
...
-  const title = metric.catalog === 'CUSTOM_METRICS' ? 'OpenSearch' : 'Prometheus';
-  const token =
-    metric.catalog === 'CUSTOM_METRICS'
-      ? '/ui/default_branding/opensearch_mark_default_mode.svg'
-      : 'tokenProperty';
-  const name = (metricName: string) => {
-    return metric.catalog === 'CUSTOM_METRICS' ? metricName : metricName.split('.')[1];
+  const name = () => {
+    if (metric.catalog === 'CUSTOM_METRICS') return metric.name;
+    else return metric.name.split('.')[1].replace(/^prometheus_/, 'p.._');
```
Comment (joshuali925) [inline public/components/metrics/sidebar/metric_name.tsx:30]: why `p..`?

File: public/components/metrics/top_menu/metrics_export_panel.tsx [hunk 5]
```diff
@@ -75,10 +76,10 @@ export const MetricsExportPanel = ({
 
   useEffect(() => {
     fetchAllvisualizationsById();
-  }, []);
+  }, [fetchAllvisualizationsById]);
```
Comment (joshuali925) [inline public/components/metrics/top_menu/metrics_export_panel.tsx:79]: wouldn't this go into a loop?
1. component mount
1. `fetchAllvisualizationsById` is called
2. `fetchAllvisualizationsById` sets visualizations metadata state
3. rerender
4. a new `fetchAllvisualizationsById` function is created
5. step 2 repeats

File: public/components/metrics/redux/slices/metrics_slice.ts [hunk 6]
```diff
@@ -123,56 +150,51 @@ export const metricSlice = createSlice({
...
 
-export const metricsStateSelector = (state) => state.metrics;
-
 export const availableMetricsSelector = (state) =>
-  state.metrics.metrics.filter(
-    (metric) => !state.metrics.selected.includes(metric.id) && !metric.recentlyCreated
-  );
+  state.metrics.metrics
+    .filter((metric) => !state.metrics.selected.includes(metric.id))
+    .filter((metric) => state.metrics.search === '' || metric.name.includes(state.metrics.search));
```
Comment (joshuali925) [inline public/components/metrics/redux/slices/metrics_slice.ts:187]: https://github.com/opensearch-project/dashboards-observability/blob/108189754cb1b98c75cc2a7e6909b990f32687b6/public/components/metrics/sidebar/search_bar.tsx#L16

this only triggers when user is typing

If user didn't type in search bar, `state.metrics.search` is undefined, and this filter statement filter out all metrics?

Comment (pjfitzgibbons) [reply public/components/metrics/redux/slices/metrics_slice.ts:188]: Updated initialState.search at top of file (ln: 42).   Default value empty string.
This change got lost in a merge.

**Comment by joshuali925** [comment]
> The function is a static value. This is a lint finding. If it's better to just have an anonymous effect, then i'll revert it.
This works fine in live code.

the function is not static, it's recreated on every render

**Review by joshuali925 — APPROVED** [review]

**Pull request merged**